File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = [0]
        def bfs(node):
            if not node:
                return 0
            left = bfs(node.left)
            right = bfs(node.right)
            res[0] = max(res[0], left + right)
            return 1 + max(left, right)
        bfs(root)
        return res[0]

File: 0543-diameter-of-binary-tree/0543-diameter-of-binary-tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
    let max = 0;
    function bfs(node) {
        if (!node) return 0;
        const left = bfs(node.left);
        const right = bfs(node.right);
        max = Math.max(max, left + right);
        return 1 + Math.max(left, right);
    }
    const _ = bfs(root);
    return max;
};